programs/order-engine/src/instructions/fill.rs

```diff
@@ -141,24 +146,121 @@ pub fn handle_fill<'c: 'info, 'info>(
                 },
             ))?;
         }
-        (Some(maker_output_mint_token_account), Some(taker_output_mint_token_account)) => {
-            token::transfer(
-                CpiContext::new(
-                    ctx.accounts.output_token_program.to_account_info(),
-                    token::Transfer {
-                        from: maker_output_mint_token_account.to_account_info(),
-                        to: taker_output_mint_token_account.to_account_info(),
-                        authority: ctx.accounts.maker.to_account_info(),
-                    },
-                ),
-                output_amount,
-            )?;
-        }
+        (Some(maker_output_mint_token_account), Some(taker_output_mint_token_account)) => transfer(
+            ctx.accounts.output_token_program.to_account_info(),
+            maker_output_mint_token_account.to_account_info(),
+            taker_output_mint_token_account.to_account_info(),
+            ctx.accounts.maker.to_account_info(),
+            ctx.accounts.output_mint.to_account_info(),
+            output_amount,
+        )?,
     }
 
     Ok(())
 }
 
+fn transfer<'info>(
+    token_program: AccountInfo<'info>,
+    from: AccountInfo<'info>,
+    to: AccountInfo<'info>,
+    authority: AccountInfo<'info>,
+    mint: AccountInfo<'info>,
+    amount: u64,
+) -> Result<()> {
+    let decimals_for_transfer_checked = if token_program.key.eq(&spl_token_2022::ID) {
+        let mint_data = mint.try_borrow_data()?;
+        let mint_state_with_extensions =
+            StateWithExtensions::<spl_token_2022::state::Mint>::unpack(&mint_data)?;
+
+        // reference https://github.com/solana-labs/solana-program-library/blob/e215d42a076d22512b60007839f3a3f216d9972d/token/program-2022/src/processor.rs#L368-L392
+        let mut has_extension_requiring_transfer_checked = false;
+        for extension_type in mint_state_with_extensions.get_extension_types()? {
```

Comment:
Another option is to only check for loss prone extension, so transfer fee non zero as it will mess up with the accounting. Then let everything else pass through